Dapper/SqlMapper.GridReader.Async.cs

```diff
@@ -104,27 +104,23 @@ private Task<IEnumerable<T>> ReadAsyncImpl<T>(Type type, bool buffered)
 
             private async Task<IEnumerable<T>> ReadBufferedAsync<T>(int index, Func<IDataReader, object> deserializer, Identity typedIdentity)
             {
-                //try
-                //{
-                var reader = (DbDataReader)this.reader;
-                List<T> buffer = new List<T>();
-                while (index == gridIndex && await reader.ReadAsync(cancel).ConfigureAwait(false))
+                try
                 {
-                    buffer.Add((T)deserializer(reader));
+                    var reader = (DbDataReader)this.reader;
+                    List<T> buffer = new List<T>();
+                    while (index == gridIndex && await reader.ReadAsync(cancel).ConfigureAwait(false))
+                    {
+                        buffer.Add((T)deserializer(reader));
+                    }
+                    return buffer;
                 }
-                if (index == gridIndex) // need to do this outside of the finally pre-C#6
+                finally // finally so that First etc progresses things even when multiple rows

```

Comment:
Yes, I think that's right; IIRC that was the code we had originally, targeting C# 6 (the language version no longer being so much of a pain point)